File: baekjoon/9663.py
# N-Queen
# 백트래킹

# 백트래킹 끝판왕... 복습 철저히

N = int(input())
isFull = [None]*N
ans = 0
#유망성 판단
def promising(depth,i) :
    if i in isFull :
        return False
    else :
        for j in range(depth) :
            if depth-j == abs(i-isFull[j]) :
                return False
        return True 

def putQueen(depth) :
    global ans
    if depth == N :
        ans += 1
        print(isFull)
        return
    for i in range(N) :
        if promising(depth,i) :
            isFull[depth] = i
            putQueen(depth+1)
            isFull[depth] = None
putQueen(0)
print(ans)

# 개쌉 야매법
# answer = [0, 1, 0, 0, 2, 10, 4, 40, 92, 352, 724, 2680, 14200, 73712, 365596]
# print(answer[int(input())])

File: baekjoon/9663.cpp
// N-Queen
// 백트래킹

#include <iostream>
#include <cstdlib>
#include <string.h>

int N, ans = 0;
int isFull[16];

bool promising(int depth, int i){
    // depth 열의 i행에 queen을 둘 수 있는지 판단
    for(int j = 0; j < N; j++){
        if(isFull[j] == i){
            return false;
        }
    }

    for(int j = 0; j < depth; j++){
        if(depth - j == std::abs(i - isFull[j])){
            return false;
        }
    }

    return true;
}

void putQueen(int depth){
    if(depth == N){
        ans += 1;
        return;
    }

    for(int i = 0; i < N; i++){
        if(promising(depth, i)){
            isFull[depth] = i;
            putQueen(depth + 1);
            isFull[depth] = -1;
        }
    }
}

int main(){

    std::ios_base::sync_with_stdio(0);
    std::cin.tie(0);

    std::cin >> N;

    memset(isFull, -1, sizeof(isFull));

    putQueen(0);

    std::cout << ans << '\n';
    
    return 0;

}